src/main/java/net/dv8tion/jda/core/utils/SessionController.java

```diff
@@ -0,0 +1,190 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.dv8tion.jda.core.utils;
+
+import net.dv8tion.jda.core.JDA;
+import net.dv8tion.jda.core.utils.tuple.Pair;
+
+/**
+ * Controls states and behaviour of one or multiple {@link net.dv8tion.jda.core.JDA JDA} instances.
+ * <br>One instance of this should be used when sharding a bot account in order to keep track of session information
+ * between shards.
+ *
+ * <p>The {@link net.dv8tion.jda.core.utils.SessionControllerAdapter SessionControllerAdapter} provides
+ * a default implementation that can be extended and overridden.
+ *
+ * <h2>States {@literal &} Behaviour</h2>
+ * <b>Identify Ratelimit Handling</b>
+ * <br>This will enable handling of (re-)connecting gateway sessions.
+ *
+ * <p><b>Global REST Ratelimit</b>
+ * <br>The global REST ratelimit is not bound to a single session and should be
+ * handled on all JDA instances. This controller will receive updates of this ratelimit through {@link #setGlobalRatelimit(long)}
+ * and should report the last ratelimit information it received through {@link #getGlobalRatelimit()}.
+ *
+ * <p><b>Gateway Provider</b>
+ * <br>This provider can be used to change the gateway retrieval (using cache, http, or static) and
+ * allows to set a custom gateway endpoint. <b>Use carefully.</b>
+ *
+ * <h2>Examples</h2>
+ *
+ * <b>Using {@link net.dv8tion.jda.core.JDABuilder JDABuilder}</b>
+ * <br>
+ * <pre><code>
+ * JDABuilder builder = new JDABuilder(AccountType.BOT).setToken(BOT_TOKEN);
+ * builder.setSessionController(new SessionControllerAdapter() {
+ *     {@literal @Override}
+ *     public void appendSession(SessionConnectNode node) {
+ *         System.out.println("[SessionController] Adding SessionConnectNode to Queue!");
+ *         super.appendSession(node);
+ *     }
+ * });
+ * builder.addEventListener(myListener);
+ * for (int i = 0; i {@literal <} 10; i++) {
+ *     builder.useSharding(i, 10).buildAsync();
+ * }
+ * </code></pre>
+ *
+ * <p><b>Using {@link net.dv8tion.jda.bot.sharding.ShardManager ShardManager}</b>
+ * <br>
+ * <pre><code>
+ * DefaultShardManagerBuilder builder = new DefaultShardManagerBuilder();
+ * builder.setToken(BOT_TOKEN);
+ * builder.setSessionController(new SessionControllerAdapter() {
+ *     {@literal @Override}
+ *     public {@literal Pair<String, Integer>} getGatewayBot(JDA api) {
+ *         return Pair.of(getGateway(), 10);
+ *     }
+ * });
+ * builder.addEventListener(myListener);
+ * builder.build();
+ * </code></pre>
+ */
+public interface SessionController
+{
+    /**
+     * The default delay (in seconds) to wait between running {@link net.dv8tion.jda.core.utils.SessionController.SessionConnectNode SessionConnectNodes}
+     */
+    int IDENTIFY_DELAY = 5;
```

Comment:
Why is this a field rather than a method? As a field of an interface it is effectively `public static final`. With my gateway I would want to replace this with 0 as the OP 2 ratelimit would be handled by the gateway.

Reply:
It is a constant from discord. If you want to remove or change the delay you have to override the worker that starts the sessions.

Reply:
0 delay:
```java
@Override
public void appendSession(SessionConnectNode node) {
    node.run(); // with try/catch
}
```

Reply:
Fair enough. I just thought it would make more sense to just extend the adapter and change that value.

Reply:
I added additional constructors for QueueWorker to make this easier.
```java
new SessionControllerAdapter()
{
    @Override
    public void runWorker()
    {
        synchronized(lock)
        {
            if (workerHandle == null)
            {
                workerHandle = new QueueWorker(0); // or 0L for millis
                workerHandle.start();
            }
        }
    }
};
```